### PR TITLE
feature: Pluggable component feature flag

### PR DESCRIPTION
**Pull request by mcandeia**
Signed-off-by: Marcos Candeia <marrcooos@gmail.com>

# Description

This PR adds feature toggle for Pluggable Components.

## Issue reference

Please reference the issue this PR will close: #5156 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/runtime/runtime.go
Comment (berndverst) [inline pkg/runtime/runtime.go:632]: IMO it should never call this function if the feature is not enabled. I would not put that within this function.

**Review by berndverst — CHANGES_REQUESTED** [review]

File: pkg/runtime/runtime.go
Comment (mcandeia) [reply pkg/runtime/runtime.go:632]: Okay, I can do that, but I chose that function because we don't have any `initRuntime` tests and I think it's a bit difficult to do because it has a lot of dependencies - and I don't believe that this is the context of this PR. Having that said, I'd rather have the test than put it inside the initRuntime, otherwise I can create an initPluggableComponents() that does this check before calling the register, at least that way it can be "testable".

wdyt @berndverst ?

Comment (berndverst) [reply pkg/runtime/runtime.go:632]: I will leave this up to dapr/dapr maintainers.

We have already proven that our feature flags work, so we don't need to specifically test for that again.

IMO your approach is less readable. If the method were called `registerPluggableComponentsIfFeatureEnabled` it makes sense - but simply `registerPluggableComponents` is weird if I am debugging a scenario where I have the feature disabled.

Thoughts?

Comment (mukundansundar) [reply pkg/runtime/runtime.go:632]: IMO registerPluggableComponents should only register any pluggable component that is there. It should not check if the feature is enabled or not. And having an initPluggableComponents make sense in a way to modularize the code as well as making it testable. And that initialization code checking if it has to start registering pluggable components if feature is enabled makes its more clear Imo.

Comment (mcandeia) [reply pkg/runtime/runtime.go:632]: >IMO your approach is less readable. If the method were called registerPluggableComponentsIfFeatureEnabled it makes sense - but simply registerPluggableComponents is weird if I am debugging a scenario where I have the feature disabled. I would not expect that method to be entered in such a case.

I would agree with you if we had a single place where featureflags are initialized/checked, which is not the case. In that sense I prefer InitPluggableComponents because it turns the entrypoint for anyone to understand if the components will be loaded or not, even because I think the initRuntime code is getting too many lines and bad to maintain, so I think discussing name when the we have untested code ends up becoming irrelevant

Comment (mcandeia) [reply pkg/runtime/runtime.go:632]: > IMO registerPluggableComponents should only register any pluggable component that is there. It should not check if the feature is enabled or not. And having an initPluggableComponents make sense in a way to modularize the code as well as making it testable. And that initialization code checking if it has to start registering pluggable components if feature is enabled makes its more clear Imo.

I agree moving to there

**Review by artursouza — APPROVED** [review]
@tmacam We can start documenting this as well.

**Comment by tmacam** [comment]
> @tmacam We can start documenting this as well.

Sure. Should this be added to `dapr-docs` directly or as part of the sandbox-projects?

**Pull request merged**
